diffkemp/simpll/InstPatternComparator.cpp

```diff
@@ -0,0 +1,655 @@
+    jumpToInst(InstL, PositionL);
+    jumpToInst(InstR, PositionR);
+
+    while (InstL != InstLE && InstR != InstRE) {
+        PositionL = &*InstL;
+        PositionR = &*InstR;
+
+        // Check whether the compared pattern ends at this instruction.
+        auto InstRMetadata = ParentPattern->MetadataMap.find(&*InstR);
+        if (InstRMetadata != ParentPattern->MetadataMap.end()
+            && InstRMetadata->second.PatternEnd)
+            return 0;
+
+        // If at the end of only one basic block, leave the rest to
+        // unconditionally connected basic blocks (if there are any).
+        if (&*InstL == TermL && &*InstR != TermR
+            && TermL->getNumSuccessors() == 1) {
+            return 1;
+        }
+        if (&*InstL != TermL && &*InstR == TermR) {
+            if (TermR->getNumSuccessors() == 1)
+                return 0;
+
+            // If no unconditional successor exists, skip to the terminator
+            // instruction.
+            InstL = InstLE;
+            --InstL;
+            continue;
+        }
+
+        // Compare current instructions with operands.
+        if (int Res = cmpOperationsWithOperands(&*InstL, &*InstR)) {
+            // Remove newly added value and input mappings.
+            eraseNewlyMapped();
+
+            // When in an instruction group, do not allow module instruction
+            // skipping.
+            if (GroupDepth > 0)
+                return Res;
+
+            // Skip the module instruction.
+            ++InstL;
+            continue;
+        }
+
+        // Register pattern instruction groups.
+        if (InstRMetadata != ParentPattern->MetadataMap.end()
+            && InstRMetadata->second.GroupStart)
+            ++GroupDepth;
+        if (InstRMetadata != ParentPattern->MetadataMap.end()
+            && InstRMetadata->second.GroupEnd)
+            --GroupDepth;
+
+        ++InstL;
+        ++InstR;
+    }
+
+    if (InstL != InstLE && InstR == InstRE)
+        return 1;
+    if (InstL == InstLE && InstR != InstRE)
+        return -1;
+
+    return 0;
+}
+
+/// Compare global values by their names, because their indexes are not
+/// expected to be the same.
+int InstPatternComparator::cmpGlobalValues(GlobalValue *L,
+                                           GlobalValue *R) const {
+    if (L->hasName() && R->hasName()) {
+        // Both values are named, compare them by names
+        auto NameL = L->getName();
+        auto NameR = R->getName();
+
+        if (namesMatch(NameL, NameR, IsLeftSide))
+            return 0;
+
+        return 1;
+    }
+
+    return L != R;
+}
+
+/// Compare a module value with a pattern value using serial numbers.
+/// Note: Parts of this function have been adapted from FunctionComparator.
+/// Therefore, LLVM licensing also applies here. See the LICENSE information
+/// in the appropriate llvm-lib subdirectory for more details.
+int InstPatternComparator::cmpValues(const Value *L, const Value *R) const {
```

Comment:
This would not be practical (although possibly doable in some way). The modified `cmpValues` needs access to `LeftSN` and `RightSN`, not just to the original result.